File: Related Topics/Two Pointers/125. Valid Palindrome.java
class Solution {
    public boolean isPalindrome(String s) {
        if (s.isEmpty()) return true;
        
        int l = 0, r = s.length() - 1;
        
        while (l < r) {
            char cl = s.charAt(l), cr = s.charAt(r);
            if (!Character.isLetterOrDigit(cl)) {
                l++;
            }
            else if (!Character.isLetterOrDigit(cr)) {
                r--;
            }
            else {
                if (Character.toLowerCase(cl) != Character.toLowerCase(cr)) {
                    return false;
                }
                l++;
                r--;
            }
        }
        
        return true;
    }
}

File: Related Topics/Two Pointers/125. Valid Palindrome.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        if not s:
            return True
        
        l, r = 0, len(s) - 1
        while l < r:
            if l < r and not s[l].isalnum():
                l += 1
            elif l < r and not s[r].isalnum():
                r -= 1
            else:
                if s[l].lower() != s[r].lower():
                    return False
                l += 1
                r -= 1
        
        return True
        